File: Exercises/rb101-rb109_small_problems/medium_1/1000lights.rb
# frozen_string_literal: true

=begin

# PROBLEM
input: positive integer
output: array of integers

explicit rules:
  - input represents total number of switches
  - input also represents total number of rounds of play
  - each switch controls exactly one light
  - initial state of lights: off
  - first round: all lights are turned on
  - in each successive round switches are toggled according to the following rules:
    - start at switch 2, then move to each multiple of 2 up to the total number of switches
    - then start at 3, then move to each multiple of 3 up to the total number of switches


# EXAMPLES
 puts switches(5) == [1, 4]
  - round 1: all lights on
  - round 2: lights 2 and 4 are off; 1, 3, 5 are on
  - round 3: lights 2, 3, and 4 are off; 1 and 5 are on
  - round 4: lights 2 and 3 are off; 1, 4 and 5 are on
  - round 5: lights 2, 3, and 5 are off; 1 and 4 are on
 puts switches(10) == [1, 4, 9]
  - round 1: all lights turned on
    - { 1: on, 2: on, 3: on, 4: on, 5: on, 6: on, 7: on, 8: on, 9: on, 10: on }
  - round 2: 2, 4, 6, 8, 10 turned off
    - { 1: on, 2: off, 3: on, 4: off, 5: on, 6: off, 7: on, 8: off, 9: on, 10: off }
  - round 3: 3, 6, 9 turned off
    - { 1: on, 2: off, 3: off, 4: off, 5: on, 6: off, 7: on, 8: off, 9: off, 10: off }
  -round 4: 4, 8 turned off
    -
 puts switches(20) ==

=end

require 'pry'

def initialize_switches(total_lights)
  switches = {}
  1.upto(total_lights) { |num| switches[num] = false }
  switches
end

def find_and_switch_lights(total_lights, switches)
  1.upto(total_lights) do |num|
    switches.each do |position, status|
      switches[position] = !status if (position % num).zero?
    end
  end
end

def collect_on_lights(switches)
  switches.each_with_object([]) do |(key, value), array|
    array << key if value
  end
end

def switch_lights(total_lights)
  switches = initialize_switches(total_lights)
  find_and_switch_lights(total_lights, switches)
  collect_on_lights(switches)
end

p switch_lights(5)  #== [1, 4]
p switch_lights(10) #== [1, 4, 9]
p switch_lights(20)
p switch_lights(1000)
p switch_lights(2000)

File: ls215/exercises/small_problems/interpretive_problem_solving/1000_lights.js
/* eslint-disable */

/*
 
### Problem Description ###
You have a bank of switches before you, numbered from 1 to n. Every switch is connected to exactly one light that is initially off. You walk down the row of switches and toggle every one of them. You walk back to the beginning of the row and start another pass. On this second pass, you toggle switches 2, 4, 6, and so on. On the third pass, you go back to the beginning again, this time toggling switches 3, 6, 9, and so on. You continue to repeat this process until you have gone through n repetitions.

Write a program that takes one argument—the total number of switches—and returns an array of the lights that are on after n repetitions.

### Problem ###
- input: integer
- output: array (of integers)
- rules:
  - explicit: 
    - input integer represents:
      - a row of switches numbered 1 to n
      - the number of rounds of toggling the switches
    - initial position of every switch is OFF (0)
    - each round traverses the entire length of the row of switches
    - each round starts at the beginning of the row
    - toggling rules:
      - first round: every swtich
      - second round: every 2nd switch
      - third round: every 3rd switch
      - fourth round: every 4th switch
      ...
    - at the end of the n rounds, return an array of the switches in the ON position
- questions:
  - input 0?
  - no input?
  - fractional input?
  - negative input?
  - string digit?

### Data Structures ###
- input: integer
- working with input: array
- output: array (of integers)

### Algorithm ###

- input = n
- create an array of n length, with each element as `false`
- iterate over the array n times:
  - starting round r is 1
  - for each iteration:
    - toggle each element's value if its index + 1 is a multiple of r(ound)
- once the array is in its final state, filter the array to have elements whose value is true
- return the resulting array


/* eslint-enable */

const INDEX_ADJUSTMENT = 1;

function lightsOn(switches) {
  let lights = Array(switches).fill(true);
  lights = toggleLights(lights);

  let lightIDS = Object.keys(lights)
                       .filter(id => lights[id]);

  return lightIDS.map(id => Number(id) + INDEX_ADJUSTMENT);
}

function toggleLights(lights) {
  for (let round = 2; round <= lights.length; round += 1) {
    lights = lights.map((light, index) => {
      if ((index + INDEX_ADJUSTMENT) % round === 0) {
        return !light;
      } else {
        return light;
      }
    });
  }
  return lights;
}

console.log(lightsOn(5));        // [1, 4]
// Detailed result of each round for `5` lights
// Round 1: all lights are on
// Round 2: lights 2 and 4 are now off;     1, 3, and 5 are on
// Round 3: lights 2, 3, and 4 are now off; 1 and 5 are on
// Round 4: lights 2 and 3 are now off;     1, 4, and 5 are on
// Round 5: lights 2, 3, and 5 are now off; 1 and 4 are on

console.log(lightsOn(100));      // [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
